library/src/main/java/com/novoda/downloadmanager/lib/DownloadThread.java

```diff
@@ -226,30 +227,30 @@ private void runInternal() {
             return;
         }
 
-        DownloadBatch currentBatch = batchRepository.retrieveBatchFor(originalDownloadInfo);
-
-        if (!downloadReadyChecker.canDownload(currentBatch)) {
-            Log.d("Download " + originalDownloadInfo.getId() + " is not ready to download: skipping");
-            return;
-        }
-
-        if (downloadStatus != DownloadStatus.RUNNING) {
-            ContentValues contentValues = new ContentValues();
-            contentValues.put(COLUMN_STATUS, DownloadStatus.RUNNING);
-            context.getContentResolver().update(originalDownloadInfo.getAllDownloadsUri(), contentValues, null, null);
-            updateBatchStatus(originalDownloadInfo.getBatchId(), originalDownloadInfo.getId());
-        }
-
-        State state = new State(originalDownloadInfo);
         PowerManager.WakeLock wakeLock = null;
         int finalStatus = DownloadStatus.UNKNOWN_ERROR;
         int numFailed = originalDownloadInfo.getNumFailed();
         String errorMsg = null;
-
-//        final NetworkPolicyManager netPolicy = NetworkPolicyManager.from(context);
-        PowerManager powerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
+        State state = new State(originalDownloadInfo);
 
         try {
+            DownloadBatch currentBatch = batchRepository.retrieveBatchFor(originalDownloadInfo);
+
+            if (!downloadReadyChecker.canDownload(currentBatch)) {
+                throw new StopRequestException(DownloadStatus.QUEUED_DUE_CLIENT_PERMISSIONS, "Cannot proceed because client denies");
```

Comment:
Using exceptions for control flow is how everything is done in the Thread atm, I think we should keep it all consistent for now and refactor it all at the same time